website/bot.py

```diff
@@ -0,0 +1,174 @@
+import tempfile
+from pathlib import Path
+
+import openai
+from django.core.files.base import ContentFile
+from django.core.files.storage import default_storage
+from dotenv import find_dotenv, load_dotenv
+from langchain.chains import ConversationalRetrievalChain
+from langchain.memory import ConversationSummaryMemory
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+from langchain_community.document_loaders import (
+    DirectoryLoader,
+    Docx2txtLoader,
+    PyPDFLoader,
+    TextLoader,
+    UnstructuredMarkdownLoader,
+)
+from langchain_community.vectorstores import FAISS
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_openai import ChatOpenAI, OpenAIEmbeddings
+from openai import OpenAI
+
+load_dotenv(find_dotenv(), override=True)
+
+
+def is_api_key_valid(api_key):
+    client = OpenAI(api_key=api_key)
+    try:
+        client.completions.create(prompt="Hello", model="gpt-3.5-turbo-instruct", max_tokens=1)
+        return True
+    except openai.APIConnectionError as e:
+        print(f"Failed to connect to OpenAI API: {e}")
+        return False
+    except openai.RateLimitError as e:
+        print(f"OpenAI API rate limit exceeded: {e}")
+        return False
+    except openai.APIError as e:
+        print(f"OpenAI API error: {e}")
+        return False
+
+
+def load_document(file_path):
+    loaders = {
+        ".pdf": PyPDFLoader,
+        ".docx": Docx2txtLoader,
+        ".txt": TextLoader,
+        ".md": UnstructuredMarkdownLoader,
+    }
+
+    file_path = Path(file_path)
+    extension = file_path.suffix
+    Loader = loaders.get(extension)
+
+    if Loader is None:
+        raise ValueError(f"Unsupported file format: {extension}")
+
+    return Loader(file_path).load()
+
+
+def load_directory(dir_path):
+    return DirectoryLoader(dir_path).load()
+
+
+def split_document(chunk_size, chunk_overlap, document):
+    text_splitter = RecursiveCharacterTextSplitter(
+        chunk_size=chunk_size,
+        chunk_overlap=chunk_overlap,
+        length_function=len,
+    )
+    return text_splitter.split_documents(document)
+
+
+def embed_documents_and_save(embed_docs):
+    db_folder_path = Path("faiss_index")
+
+    embeddings = OpenAIEmbeddings(model="text-embedding-3-small")
+
+    # Temporary directory for local operations
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_db_path = Path(temp_dir) / db_folder_path
+        temp_db_path.mkdir(parents=True, exist_ok=True)
+
+        # Check if the folder exists in the storage system and download files
+        if default_storage.exists(str(db_folder_path)) and default_storage.listdir(
```

Comment:
It seems you could use an additional variable to store and reuse `str(db_folder_path)`